hazelcast/src/main/java/com/hazelcast/query/impl/CompositeConverter.java

```diff
@@ -89,18 +89,18 @@ public Comparable convert(Comparable value) {
         }
         CompositeValue compositeValue = (CompositeValue) value;
 
-        // Components are converted in-place to reduce littering. Composite
-        // values are never exposed to user code, so that's not a problem to
-        // mutate them.
         Comparable[] components = compositeValue.getComponents();
+        Comparable[] converted = new Comparable[components.length];
         for (int i = 0; i < components.length; ++i) {
             Comparable component = components[i];
-            if (component != NULL && component != NEGATIVE_INFINITY && component != POSITIVE_INFINITY) {
-                components[i] = converters[i].convert(component);
+            if (component == NULL || component == NEGATIVE_INFINITY || component == POSITIVE_INFINITY) {
+                converted[i] = component;
```

Comment:
This branch seems like not covered with tests. Would you mind adding a test to `CompositeConverterTest`?

Reply:
Added a test to cover special values conversion.